core/org.wso2.carbon.ndatasource.rdbms/src/main/java/org/wso2/carbon/ndatasource/rdbms/CorrelationLogInterceptor.java

```diff
@@ -0,0 +1,219 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.ndatasource.rdbms;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.tomcat.jdbc.pool.interceptor.AbstractQueryReport;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.sql.DatabaseMetaData;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * Time-Logging interceptor for JDBC pool.
+ * Logs the time taken to execute the query in each pool-ed connection.
+ */
+
+public class CorrelationLogInterceptor extends AbstractQueryReport {
+
+    private static final Log correlationLog = LogFactory.getLog("CORRELATION_LOG");
+    private static Log log = LogFactory.getLog(CorrelationLogInterceptor.class);
+
+    private static final String CORRELATION_LOG_TIME_TAKEN_KEY = "delta";
+    private static final String CORRELATION_LOG_TIME_TAKEN_UNIT = "ms";
+    private static final String CORRELATION_LOG_CALL_TYPE_KEY = "callType";
+    private static final String CORRELATION_LOG_CALL_TYPE_VALUE = "jdbc";
+    private static final String CORRELATION_LOG_START_TIME_KEY = "startTime";
+    private static final String CORRELATION_LOG_METHOD_NAME_KEY = "methodName";
+    private static final String CORRELATION_LOG_QUERY_KEY = "query";
+    private static final String CORRELATION_LOG_CONNECTION_URL_KEY = "connectionUrl";
+    private static final String CORRELATION_LOG_SEPARATOR = " | ";
+    private static final String CORRELATION_LOG_SYSTEM_PROPERTY = "enableCorrelationLogs";
+
+    @Override
+    public void closeInvoked() {
+
+    }
+
+    @Override
+    protected void prepareStatement(String s, long l) {
+
+    }
+
+    @Override
+    protected void prepareCall(String s, long l) {
+
+    }
+
+    @Override
+    public Object createStatement(Object proxy, Method method, Object[] args, Object statement, long time) {
+
+        try {
+            if (Boolean.parseBoolean(System.getProperty(CORRELATION_LOG_SYSTEM_PROPERTY))) {
+                return invokeProxy(method, args, statement, time);
+            } else {
+                return statement;
+            }
+
+        } catch (Exception e) {
+            log.warn("Unable to create statement proxy for slow query report.", e);
+            return statement;
+        }
+    }
+
+    private Object invokeProxy(Method method, Object[] args, Object statement, long time) throws Exception {
+
+        String name = method.getName();
+        String sql = null;
+        Constructor<?> constructor = null;
+
+        if (this.compare("prepareStatement", name)) {
+            sql = (String) args[0];
+            constructor = this.getConstructor(1, PreparedStatement.class);
+            if (sql != null) {
+                this.prepareStatement(sql, time);
```

Comment:
This is just an empty method we are calling. What is the purpose of calling this ?